### PR TITLE
feat(client): add dynamo_chat transport + routed_experts to renderer generate

### PR DESCRIPTION
**Pull request by biswapanda**
## Description

Adds a `dynamo_chat` transport to the renderer-based `generate()` client so it can run against NVIDIA Dynamo, which serves no `/inference/v1/generate` route. Selected per-call via `transport=`; defaults to the existing vLLM path, so behavior is unchanged unless opted in.

Two transports:

- **`vllm_generate` (default)**: unchanged — `messages → render_ids() → POST /inference/v1/generate → parse_response()` (vLLM TITO surface).
- **`dynamo_chat`**: `messages → render_ids() → POST /v1/chat/completions` with `nvext.token_data` (pre-tokenized prompt) + `nvext.extra_fields=["engine_data"]`. Completion token IDs and logprobs are read back from `nvext.engine_data`.

### Dynamo wire shape (`_post_dynamo_chat`)

Mirrors the verifiers token client so the payload is identical whether a rollout goes through the token client or the renderer client. `nvext.token_data` (Dynamo skips tokenization when present); `cache_salt` → `nvext.cache_salt`, `priority` → `nvext.agent_hints.priority`; a single placeholder user message; sampling remap (`max_tokens` → `max_completion_tokens`, `logprobs=N` → `logprobs=true` + `top_logprobs=N`); passthrough fields ride the Dynamo allowlist. Tools are baked into `token_data` by the renderer (not sent on the wire).

### routed_experts (MoE expert replay) — now surfaced on dynamo_chat

(Supersedes the earlier "routed_experts intentionally NOT surfaced" note — it now is.) `parse` reads routed_experts from `nvext.routed_experts` (or `nvext.engine_data.routed_experts`) and maps it to the downstream `RoutedExpertsPayload` `{data, shape, start, dtype}`. The Dynamo worker returns full-sequence routing with `start=0`; the renderer row-trims the leading prompt rows **only when the caller explicitly sets `routed_experts_prompt_start`** — a first-turn request with no caller start stays full-sequence with `start=0` (no phantom prefix). Completion logprobs prefer `nvext.engine_data.completion_logprobs` (the same authoritative source as the engine token IDs) over the chat echo; a present-but-empty engine list is authoritative and does not fall back to chat.

### Other

- Public `RendererTransport = Literal["vllm_generate", "dynamo_chat"]` alias. A present-but-empty `completion_token_ids` is a valid zero-token completion; only a fully absent field raises. Multimodal renderers raise `NotImplementedError` on `dynamo_chat` (vLLM path / token-client TITO remain available for VLMs).

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Review

Codex adversarial review: **SIGN-OFF** (F1/F2/F3 + the N1 logprob-presence finding resolved; head `5f2a914`). All review threads resolved.

## Testing

`tests/test_client.py` covers the Dynamo request body shape (priority/detokenize/sampling remap), routed_experts parse + row-trim (explicit `prompt_start` vs first-turn full-sequence), engine-logprob preference incl. present-but-empty, and missing/empty completion IDs.

---

> [!NOTE]
> **Medium Risk**
> New Dynamo wire/parse path affects RL-critical completion IDs, logprobs, and MoE `routed_experts`; strict runtime errors and no Dynamo multimodal are new failure modes for opted-in rollouts.
> 
> **Overview**
> Adds a per-call **`transport`** parameter to **`generate()`** (`"vllm"` default, `"dynamo"` opt-in). The existing vLLM TITO flow is moved into **`_VllmGenerateTransport`**; behavior stays the same when **`transport`** is omitted.
> 
> **Dynamo** uses **`_DynamoChatTransport`**: pre-tokenized prompts go to **`POST /v1/chat/completions`** via **`nvext.token_data`**, with **`cache_salt`**, **`priority`**, and **`routed_experts_prompt_start`** mapped into **`nvext`** and vLLM-only sampling keys dropped. Responses read **`nvext.engine_data`** for completion IDs and logprobs (not chat echo), normalize **`routed_experts`**, keep large blobs as zero-copy **`memoryview`**, and optionally client-trim prompt rows when an older worker returns full-sequence routing.
> 
> Multimodal on Dynamo raises **`NotImplementedError`**; missing engine completion IDs or logprob length mismatches raise **`RuntimeError`**. Tests cover wire shape, **`nvext`** merge, and parse edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 57846ecf216104f9ad1d9cf01541c2400744325e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add `dynamo` transport and routed-experts support to `generate()` in the renderer client
> - Adds a `transport` parameter to `generate()` in [renderers/client.py](https://github.com/PrimeIntellect-ai/renderers/pull/79/files#diff-4eceef280041aa5c33da06241f0ca20ad18ca3b7a395767ebc916e9a46ff2f33), defaulting to `'vllm'` (existing `/inference/v1/generate` path); passing `'dynamo'` routes to OpenAI-compatible `/v1/chat/completions` with NVIDIA Dynamo `nvext` fields.
> - Introduces a `_Transport` ABC with `_VllmGenerateTransport` and `_DynamoChatTransport` implementations; each handles body construction, POST, and response normalization into a common `_WireResult`.
> - The Dynamo transport maps sampling params (dropping vLLM-only keys), moves `cache_salt`/`priority` into `nvext`, and prefers `engine_data` fields over chat-echo fields when parsing responses.
> - Adds client-side trimming of base64-encoded `routed_experts` via `_trim_dynamo_routed_experts` when `routed_experts_prompt_start` is set and the worker has not already trimmed.
> - Risk: `generate()` with `transport='dynamo'` raises `RuntimeError` on missing `completion_token_ids` or logprob/token-ID length mismatches, where the vLLM path does not.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 57846ec.</sup>
> 
>

### CODE REVIEW

File: renderers/client.py [hunk 1]
```diff
@@ -107,6 +122,326 @@ async def _resolve_max_prompt_len(client: AsyncOpenAI, model: str) -> int | None
...
+            raise RuntimeError(
+                "dynamo_chat response carried no completion token IDs "
+                "(expected nvext.engine_data.completion_token_ids)."
+            )
+        completion_ids = list(completion_ids or [])
+
+        logprobs = _flatten_chat_logprobs(choice)
+        if not logprobs:
+            # Dynamo-only fallback when chat logprobs are absent.
+            logprobs = [float(x) for x in engine.get("completion_logprobs") or []]
```
[the remark this answers is not in the document]
Comment (biswapanda) [reply renderers/client.py]: Fixed in 51d9154: parse() now prefers `engine_data.completion_logprobs` (the same authoritative source as the engine `completion_token_ids`) over the choices[0] chat echo, falling back to chat only when the engine channel carries none — so ids and logprobs stay aligned.

File: renderers/client.py [hunk 2]
```diff
@@ -107,6 +123,427 @@ async def _resolve_max_prompt_len(client: AsyncOpenAI, model: str) -> int | None
...
+            finish_reason=choice.get("finish_reason"),
+        )
+
+
+class _DynamoChatTransport(_Transport):
+    """NVIDIA Dynamo: ``POST /v1/chat/completions`` with the nvext envelope.
+
+    Dynamo has no /inference/v1/generate route. ``nvext.token_data`` carries
+    the pre-tokenized prompt (Dynamo skips tokenization when present) and
+    ``extra_fields=["engine_data", "routed_experts"]`` opts into the
```
Comment (AmeenP) [inline renderers/client.py]: we don't request routed experts anymore here right?

File: renderers/client.py [hunk 3]
```diff
@@ -107,6 +123,427 @@ async def _resolve_max_prompt_len(client: AsyncOpenAI, model: str) -> int | None
...
+    if offset is None:
+        return
+    # Worker already trimmed engine-side (stamped start > 0) — don't double-trim.
+    if int(routed.get("start") or 0) != 0:
+        return
+    offset = max(0, min(int(offset), int(shape[0])))
+    if offset == 0:
+        return
+
+    itemsize = _ROUTED_EXPERTS_ITEMSIZE.get(routed.get("dtype", "uint8"), 1)
```
Comment (AmeenP) [inline renderers/client.py]: might be better to hard fail instead of falling back to 1

File: renderers/client.py [hunk 4]
```diff
@@ -107,6 +123,427 @@ async def _resolve_max_prompt_len(client: AsyncOpenAI, model: str) -> int | None
...
+        if cache_salt is not None:
+            nvext["cache_salt"] = cache_salt
+        if priority is not None:
+            agent_hints = dict(nvext.get("agent_hints") or {})
+            agent_hints["priority"] = priority
+            nvext["agent_hints"] = agent_hints
+        # routed_experts_prompt_start rides nvext (Dynamo rejects unknown
+        # top-level chat fields). The worker applies it to SamplingParams so vLLM
+        # trims the leading prompt rows engine-side and stamps the payload's
+        # `start` — the client-side trim then no-ops (see _trim_dynamo_routed_experts).
```
Comment (AmeenP) [inline renderers/client.py]: some of the comments are very large and duplicated, can we minimize to needed ones

File: renderers/client.py (hunk 2 above)
Comment (biswapanda) [reply renderers/client.py]: Correct — only  is requested. The worker nests `routed_experts` inside `engine_data.routed_experts`, so requesting the dedicated `"routed_experts"` extra field separately would duplicate the (potentially large) base64 blob on the wire. Updated the class docstring to reflect this accurately.

File: renderers/client.py (hunk 3 above)
Comment (biswapanda) [reply renderers/client.py]: Agreed. Fixed in 57846ec: now raises `RuntimeError("unknown routed_experts dtype {dtype!r}; expected one of ...")` instead of silently using `itemsize=1`. Test added for the `float16` unknown-dtype case.

File: renderers/client.py (hunk 4 above)
Comment (biswapanda) [reply renderers/client.py]: Done in 57846ec — trimmed the verbose/repeated comment blocks in `_build_body` and `_DynamoChatTransport.post()`/`parse()` to one-liners. Also corrected an inaccurate class docstring that still mentioned `extra_fields=["engine_data", "routed_experts"]`.
